File: 2021_summer_algorithm_study_basic/using_python/2180.py
"""
input :
3
2 0
1 2
0 3

output :
5
"""
import sys
sys.stdin = open("input.txt", "r")
input = sys.stdin.readline
MAX = 10 ** 9
MOD = 40000


def comp(a, b):
    if a == 0:
        return MAX
    return b / a


if __name__ == "__main__":
    N = int(input())
    ab = []
    for _ in range(N):
        a, b = map(int, input().rstrip().split())
        ab.append([a, b])

    ab.sort(key=lambda ab: comp(ab[0], ab[1]))
    print(ab)
    tot = 0
    for a, b in ab:
        tot += ((a * tot) % MOD + b) % MOD
    print(tot % MOD)

File: 2021_summer_algorithm_study_basic/using_cpp/2180.cpp
/*
input :
3
2 0
1 2
0 3

output :
5
*/

#include <iostream>
#include <algorithm>
#include <vector>
#define MOD (40000)
using namespace std;
using ll = long long;

typedef struct ab{
    double a, b;
};

// 전역 변수 선언
int N;
vector<ab> ab_vec;

bool comp(ab struct1, ab struct2){
    if (struct1.a == 0) return false;
    if (struct2.a == 0) return true;
    return struct1.b / struct1.a < struct2.b / struct2.a;
}

int main(){
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    freopen("input.txt", "r", stdin);

    cin >> N;
    while (N--){
        double a, b;
        cin >> a >> b;
        ab_vec.push_back({a, b});
    }
    sort(ab_vec.begin(), ab_vec.end(), comp);

    ll tot = 0;
    for (ab struct1: ab_vec){
        tot += (((ll)struct1.a * tot) % MOD + (ll)struct1.b) % MOD;
    }
    cout << tot % MOD;

    return 0;
}